### PR TITLE
Provide option to auto-install DVC if unavailable

### PR DESCRIPTION
**Pull request by mattseddon**
# 1/2 `main` <- this <- #2990 

This PR adds the option to auto-install DVC from the Setup webview if a Python interpreter can be found on the system.

Before shipping this change I will need to integrate the Setup webview with the current process of showing toast pop-ups when initialization fails.

#### Old

https://user-images.githubusercontent.com/37993418/208381006-502b619c-9b25-4433-b035-809d322863fa.mov

#### Current

https://user-images.githubusercontent.com/37993418/209023548-8413ded9-772e-4e8e-b206-1abfbbc6bd7c.mov

~Asked for some early feedback in #2925~

### CODE REVIEW

File: extension/src/experiments/workspace.ts [hunk 1]
```diff
@@ -329,9 +329,19 @@ export class WorkspaceExperiments extends BaseWorkspaceWebviews<
...
+
+    if (repositories.some(repository => repository.getHasData())) {
+      return true
+    }
+
+    if (repositories.some(repository => repository.getHasData() === false)) {
+      return false
+    }
+
+    return allLoading
```
Comment (mattseddon) [inline extension/src/experiments/workspace.ts:344]: [F] Needed for the edge case that the current project has data but CLI needs to be installed. I.e [this](https://github.com/iterative/vscode-dvc/pull/2944/files#diff-f1285e681b370f15ccca1f8455d2009a0dcde397464d0e5092282a932ad4962cR78)

File: extension/src/extension.ts [hunk 2]
```diff
@@ -241,6 +241,7 @@ export class Extension extends Disposable implements IExtension {
       this.config.onDidChangeExecutionDetails(async () => {
         const stopWatch = new StopWatch()
         try {
+          this.changeSetupStep()
```
Comment (mattseddon) [inline extension/src/extension.ts:244]: [F] Will update the details displayed on the Setup screen for where the CLI will be installed.

File: extension/src/setup/index.ts [hunk 3]
```diff
@@ -60,17 +60,21 @@ export class Setup extends BaseRepository<TSetupData> {
       return
     }
 
+    const pythonBinPath = await findPythonBinForInstall()
+
     this.webviewMessages.sendWebviewMessage(
       cliAccessible,
       projectInitialized,
+      isPythonExtensionInstalled(),
+      getPythonBinDisplayText(pythonBinPath),
```
Comment (mattseddon) [inline extension/src/setup/index.ts:69]: [I] Move this

File: extension/src/setup/webview/messages.ts [hunk 4]
```diff
@@ -21,19 +27,53 @@ export class WebviewMessages {
...
-      projectInitialized
+      isPythonExtensionInstalled,
+      projectInitialized,
+      pythonBinPath
     })
   }
 
   public handleMessageFromWebview(message: MessageFromWebview) {
     if (message.type === MessageFromWebviewType.INITIALIZE_PROJECT) {
       return this.initializeProject()
```
Comment (mattseddon) [inline extension/src/setup/webview/messages.ts:45]: [I] This still needs analytics attached. Probably should be done in the extension.

File: webview/src/setup/components/CliUnavailable.tsx [hunk 5]
```diff
@@ -1,16 +1,89 @@
...
+}
+
+const OfferToInstall: React.FC<{
+  children: React.ReactNode
+  pythonBinPath: string
+  installDvc: () => void
+}> = ({ installDvc, pythonBinPath, children }) => (
+  <div>
+    <p>
+      {`DVC & DVCLive can be auto-installed as packages with ${pythonBinPath}`}
```
Comment (julieg18) [inline webview/src/setup/components/CliUnavailable.tsx:20]: Technically, I don't think we have to wrap the entire string with curly braces: 
```suggestion
      DVC & DVCLive can be auto-installed as packages with {pythonBinPath}
```

File: extension/src/setup/autoInstall.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import { getPythonExecutionDetails } from '../extensions/python'
```
Comment (julieg18) [inline extension/src/setup/autoInstall.ts:1]: When I install DVC but the project isn't initialized, it takes me straight to the "Your project has no data" page. Shouldn't we be showing a message saying you need to initialize DVC?

Comment (julieg18) [reply extension/src/setup/autoInstall.ts:1]: Ah, wait, I'm seeing this behavior on main as well.

**Review by julieg18 — APPROVED** [review]
Great work!

File: extension/src/setup/autoInstall.ts (hunk 6 above)
Comment (mattseddon) [reply extension/src/setup/autoInstall.ts:1]: 😱

Comment (mattseddon) [reply extension/src/setup/autoInstall.ts:1]: @julieg18 I cannot recreate this:

https://user-images.githubusercontent.com/37993418/209230803-d7414ba9-6928-4f2f-8575-9be905d56438.mov

Are you sure the workspace that you were testing against didn't contain a DVC project?

Comment (julieg18) [reply extension/src/setup/autoInstall.ts:1]: > @julieg18 I cannot recreate this:

Probably an issue on my end. 

> Are you sure the workspace that you were testing against didn't contain a DVC project?

Yes, pretty sure. It shows no file when I try `find . -name .dvc`. I'm getting this error when the extension or I try running a dvc command:

```
ERROR: /Users/juliegalvan is not a git repository
```

Comment (mattseddon) [reply extension/src/setup/autoInstall.ts:1]: that scenario is covered by #2947

Comment (mattseddon) [reply extension/src/setup/autoInstall.ts:1]: for anyone that arrives here: debugged on a call with @julieg18 

there was a `.dvc` folder under `/Users/juliegalvan` which was causing some very unexpected/confusing behaviour. After deleting that folder everything was resolved. An issue will be raised in the DVC repo to discuss

Comment (julieg18) [reply extension/src/setup/autoInstall.ts:1]: Issue created! https://github.com/iterative/dvc/issues/8731

**Pull request merged**